src/react/reconcilation.js

```diff
@@ -277,41 +285,163 @@ export class Reconciler {
+    const contextConsumer = Get(this.realm, typeValue, "context");
+    invariant(contextConsumer instanceof ObjectValue || contextConsumer instanceof AbstractObjectValue);
+    let lastValueProp = getProperty(this.realm, contextConsumer, "currentValue");
+
+    const setContextCurrentValue = value => {
+      if (value instanceof Value) {
+        // update the currentValue
+        setProperty(this.realm, contextConsumer, "currentValue", value);
+      }
     };
+    // if we have a value prop, set it
+    if (propsValue instanceof ObjectValue || propsValue instanceof AbstractObjectValue) {
+      let valueProp = Get(this.realm, propsValue, "value");
+      setContextCurrentValue(valueProp);
+    }
+
+    newBranchState.addContext(contextConsumer);
+    if (this.componentTreeConfig.firstRenderOnly) {
+      if (propsValue instanceof ObjectValue) {
+        this._resolveReactElementHostChildren(
+          componentType,
+          reactElement,
+          context,
+          branchStatus,
+          newBranchState,
+          evaluatedChildNode
+        );
+        if (branchState !== null) {
+          branchState.mergeBranchState(newBranchState);
+        }
+        setContextCurrentValue(lastValueProp);
+        // if we no dead ends, we know the rest of the tree and can safely remove the provider
+        if (this.componentTreeState.deadEnds === 0) {
+          let childrenValue = getProperty(this.realm, propsValue, "children");
+          evaluatedChildNode.status = "INLINED";
+          this.statistics.inlinedComponents++;
+          return childrenValue;
+        }
+        return reactElement;
+      }
+    }
+    let children = this._resolveReactElementHostChildren(
+      componentType,
+      reactElement,
+      context,
+      branchStatus,
+      newBranchState,
+      evaluatedChildNode
+    );
+    if (branchState !== null) {
+      branchState.mergeBranchState(newBranchState);
+    }
+    setContextCurrentValue(lastValueProp);
+    return children;
+  }
+
+  _resolveContextConsumerComponent(
+    componentType: Value,
+    reactElement: ObjectValue,
+    branchState: BranchState | null,
+    evaluatedNode: ReactEvaluatedNode
+  ): Value | void {
+    let typeValue = Get(this.realm, reactElement, "type");
+    let propsValue = Get(this.realm, reactElement, "props");
+    let evaluatedChildNode = createReactEvaluatedNode("RENDER_PROPS", "Context.Consumer");
+    evaluatedNode.children.push(evaluatedChildNode);
 
-    if (props instanceof ObjectValue || props instanceof AbstractObjectValue) {
+    if (propsValue instanceof ObjectValue || propsValue instanceof AbstractObjectValue) {
+      // get the "render" prop child off the instance
+      let renderProp = Get(this.realm, propsValue, "children");
+      if (renderProp instanceof ECMAScriptSourceFunctionValue && renderProp.$Call) {
+        if (this.componentTreeConfig.firstRenderOnly) {
```

Comment:
Why is this for first render only? Can't we always fold context if we know its provider?

Reply:
No, because on update render, we need keep the children closure otherwise updates to it won't work.